File: .claude/templates/debug_script.m
% Debug script for investigating [specific issue/behavior]
%
% Target module: mrst_simulation_scripts/[module_name].m
% Function: [function_name]
% Date: [YYYY-MM-DD]
% Author: [Name]
%
% Issue Description:
%   - What behavior are we investigating?
%   - What error or unexpected result occurs?
%   - Under what conditions?
%
% Hypothesis:
%   - What might be causing the issue?

% ----------------------------------------
% Step 1 – Setup Debug Environment
% ----------------------------------------

% Clear workspace but keep breakpoints
clear variables; close all; clc;

% Add paths
addpath('..');
addpath('../mrst_simulation_scripts');

% Start diary for logging
diary_file = sprintf('debug/debug_log_%s.txt', datestr(now, 'yyyymmdd_HHMMSS'));
diary(diary_file);

fprintf('=====================================\n');
fprintf('DEBUG INVESTIGATION STARTED\n');
fprintf('Time: %s\n', datestr(now));
fprintf('=====================================\n\n');

% ----------------------------------------
% Step 2 – Load Test Data
% ----------------------------------------

fprintf('--- Loading Test Data ---\n');

try
    % Option 1: Load saved problematic data
    load('debug/problem_data.mat');
    fprintf('✅ Loaded saved test data\n');
    whos  % Display loaded variables
catch
    % Option 2: Generate synthetic test case
    fprintf('⚠️  No saved data found, generating synthetic test case\n');
    
    % Example: Create test grid
    nx = 20; ny = 20; nz = 5;
    test_data.grid = cartGrid([nx, ny, nz], [100, 100, 20]);
    test_data.rock = makeRock(test_data.grid, 100*milli*darcy, 0.2);
end

% Examine data characteristics
fprintf('\nData characteristics:\n');
if exist('test_data', 'var') && isstruct(test_data)
    fields = fieldnames(test_data);
    for i = 1:length(fields)
        field = fields{i};
        value = test_data.(field);
        fprintf('  %s: ', field);
        if isnumeric(value)
            fprintf('size=%s, class=%s, range=[%.4g, %.4g]\n', ...
                    mat2str(size(value)), class(value), ...
                    min(value(:)), max(value(:)));
        else
            fprintf('class=%s\n', class(value));
        end
    end
end

% ----------------------------------------
% Step 3 – Reproduce Issue
% ----------------------------------------

fprintf('\n--- Reproducing Issue ---\n');

% Test Case 1: Normal operation
fprintf('\nTest 1: Normal operation\n');
try
    % Run function with normal inputs
    normal_input = test_data;
    tic;
    result1 = suspicious_function(normal_input);
    elapsed1 = toc;
    
    fprintf('✅ Normal case passed (elapsed: %.3f s)\n', elapsed1);
    fprintf('   Result size: %s\n', mat2str(size(result1)));
catch ME
    fprintf('❌ Normal case failed: %s\n', ME.message);
    fprintf('   Error ID: %s\n', ME.identifier);
    fprintf('   Stack trace:\n');
    for i = 1:length(ME.stack)
        fprintf('     %s (line %d)\n', ME.stack(i).name, ME.stack(i).line);
    end
end

% Test Case 2: Edge case that triggers issue
fprintf('\nTest 2: Edge case\n');
try
    % Modify data to trigger issue
    edge_case_data = test_data;
    % Example: Set extreme values
    if isfield(edge_case_data, 'rock') && isfield(edge_case_data.rock, 'perm')
        edge_case_data.rock.perm(1) = inf;  % Infinite permeability
    end
    
    tic;
    result2 = suspicious_function(edge_case_data);
    elapsed2 = toc;
    
    fprintf('✅ Edge case passed (elapsed: %.3f s)\n', elapsed2);
catch ME
    fprintf('❌ Edge case failed (EXPECTED): %s\n', ME.message);
    fprintf('   Error location: %s, line %d\n', ...
            ME.stack(1).name, ME.stack(1).line);
end

% ----------------------------------------
% Step 4 – Deep Dive Analysis
% ----------------------------------------

fprintf('\n--- Deep Dive Analysis ---\n');

% Enable verbose mode if available
if exist('suspicious_function', 'file')
    % Check function internals
    fprintf('\nAnalyzing function behavior:\n');
    
    % Test with minimal input
    minimal_input = test_data;
    if isfield(minimal_input, 'grid')
        minimal_input.grid = cartGrid([2, 2, 1], [10, 10, 1]);
    end
    
    % Run with debugging output
    dbg_mode = true;
    try
        fprintf('Running with minimal input and debug mode...\n');
        result_debug = suspicious_function(minimal_input, 'debug', dbg_mode);
    catch ME
        fprintf('Debug run failed at: %s\n', ME.message);
    end
end

% Check for common issues
fprintf('\nChecking for common issues:\n');

% Check 1: NaN or Inf values
if exist('test_data', 'var') && isstruct(test_data)
    fields = fieldnames(test_data);
    for i = 1:length(fields)
        if isnumeric(test_data.(fields{i}))
            data = test_data.(fields{i});
            nan_count = sum(isnan(data(:)));
            inf_count = sum(isinf(data(:)));
            
            if nan_count > 0 || inf_count > 0
                fprintf('  ⚠️  Field "%s": %d NaN, %d Inf values\n', ...
                        fields{i}, nan_count, inf_count);
            end
        end
    end
end

% ----------------------------------------
% Step 5 – Visualization
% ----------------------------------------

fprintf('\n--- Creating Debug Visualizations ---\n');

% Create figure for debug plots
figure('Name', 'Debug Analysis', 'Position', [100, 100, 1200, 800]);

% Plot 1: Data visualization
subplot(2, 2, 1);
if exist('test_data', 'var') && isfield(test_data, 'grid')
    plotGrid(test_data.grid, 'FaceAlpha', 0.3);
    title('Grid Structure');
    view(3); axis tight;
else
    % Generic data plot
    plot(1:10, rand(1,10));
    title('Test Data Sample');
end

% Plot 2: Error analysis
subplot(2, 2, 2);
% Add problem-specific visualization
title('Error Analysis');

% Plot 3: Performance metrics
subplot(2, 2, 3);
if exist('elapsed1', 'var') && exist('elapsed2', 'var')
    bar([elapsed1, elapsed2]);
    set(gca, 'XTickLabel', {'Normal', 'Edge Case'});
    ylabel('Time (s)');
    title('Performance Comparison');
end

% Plot 4: Custom analysis
subplot(2, 2, 4);
% Add custom visualization
title('Custom Analysis');

% Save figure
saveas(gcf, 'debug/debug_analysis.png');
fprintf('✅ Saved visualization to debug/debug_analysis.png\n');

% ----------------------------------------
% Step 6 – Memory and Performance Analysis
% ----------------------------------------

fprintf('\n--- Memory and Performance ---\n');

% Check memory usage
mem_info = memory;
fprintf('Memory usage:\n');
fprintf('  Available: %.2f GB\n', mem_info.MemAvailableAllArrays/1e9);
fprintf('  Used by MATLAB: %.2f GB\n', mem_info.MemUsedMATLAB/1e9);

% Profile if needed
if false  % Set to true to enable profiling
    profile on;
    suspicious_function(test_data);
    profile off;
    profile_data = profile('info');
    % Analyze profile data
end

% ----------------------------------------
% Step 7 – Findings and Conclusions
% ----------------------------------------

fprintf('\n=====================================\n');
fprintf('FINDINGS AND CONCLUSIONS\n');
fprintf('=====================================\n');

findings = struct();
findings.timestamp = datestr(now);
findings.issue = 'Description of the issue';
findings.root_cause = 'Identified root cause';
findings.conditions = {
    'Condition 1 that triggers issue';
    'Condition 2 that triggers issue'
};
findings.proposed_fix = 'Suggested solution';

fprintf('\n📋 Summary:\n');
fprintf('1. Issue occurs when:\n');
for i = 1:length(findings.conditions)
    fprintf('   - %s\n', findings.conditions{i});
end
fprintf('2. Root cause: %s\n', findings.root_cause);
fprintf('3. Proposed fix: %s\n', findings.proposed_fix);

% Save findings
save('debug/findings.mat', 'findings');
fprintf('\n✅ Findings saved to debug/findings.mat\n');

% ----------------------------------------
% Step 8 – Cleanup
% ----------------------------------------

fprintf('\n--- Debug Session Completed ---\n');
fprintf('Outputs saved:\n');
fprintf('  - Log: %s\n', diary_file);
fprintf('  - Plots: debug/debug_analysis.png\n');
fprintf('  - Findings: debug/findings.mat\n');

% Close diary
diary off;

File: .claude/templates/debug_script.py
"""
Debug script for investigating [specific issue/behavior].

Target module: src/[module_name].py
Function: [function_name]
Date: [YYYY-MM-DD]
Author: [Name]

Issue Description:
- What behavior are we investigating?
- What error or unexpected result occurs?
- Under what conditions?

Hypothesis:
- What might be causing the issue?
"""

# ----------------------------------------
# Step 1 – Setup Debug Environment
# ----------------------------------------

# Substep 1.1 – Configure paths and imports ______________________
import sys
import os
sys.path.insert(0, os.path.abspath('..'))

# Debug utilities
import logging
import traceback
from datetime import datetime
import json

# Visualization
import matplotlib.pyplot as plt
import numpy as np

# Target module imports
try:
    from src.module_name import suspicious_function, helper_function
    print("✅ Successfully imported target functions")
except ImportError as e:
    print(f"❌ Import failed: {e}")
    traceback.print_exc()

# ----------------------------------------
# Step 2 – Configure Debug Logging
# ----------------------------------------

# Create debug logger
debug_logger = logging.getLogger('debug')
debug_logger.setLevel(logging.DEBUG)

# File handler for debug log
fh = logging.FileHandler(f'debug/debug_log_{datetime.now():%Y%m%d_%H%M%S}.log')
fh.setLevel(logging.DEBUG)

# Console handler
ch = logging.StreamHandler()
ch.setLevel(logging.INFO)

# Formatter
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
fh.setFormatter(formatter)
ch.setFormatter(formatter)

debug_logger.addHandler(fh)
debug_logger.addHandler(ch)

# ----------------------------------------
# Step 3 – Load Test Data
# ----------------------------------------

print("\n" + "="*50)
print("STARTING DEBUG INVESTIGATION")
print("="*50)

# Load problematic data
try:
    # Option 1: Load from file
    test_data = np.load('debug/problem_data.npy')
    print(f"✅ Loaded test data: shape={test_data.shape}, dtype={test_data.dtype}")
except:
    # Option 2: Generate synthetic test case
    print("⚠️  No saved data found, generating synthetic test case")
    test_data = np.random.randn(100, 50)

# Examine data characteristics
print(f"\nData statistics:")
print(f"  Min: {np.min(test_data):.4f}")
print(f"  Max: {np.max(test_data):.4f}")
print(f"  Mean: {np.mean(test_data):.4f}")
print(f"  Std: {np.std(test_data):.4f}")
print(f"  NaN count: {np.sum(np.isnan(test_data))}")

# ----------------------------------------
# Step 4 – Reproduce Issue
# ----------------------------------------

print("\n" + "-"*50)
print("REPRODUCING ISSUE")
print("-"*50)

# Test Case 1: Normal operation
print("\nTest 1: Normal operation")
try:
    result1 = suspicious_function(test_data[:10])
    print(f"✅ Normal case passed: result shape = {result1.shape}")
    debug_logger.info(f"Normal case result: {result1}")
except Exception as e:
    print(f"❌ Normal case failed: {e}")
    debug_logger.error(f"Normal case error: {e}", exc_info=True)

# Test Case 2: Edge case that triggers issue
print("\nTest 2: Edge case")
try:
    # Modify data to trigger issue
    edge_case_data = test_data.copy()
    edge_case_data[0, 0] = np.inf  # Example modification
    
    result2 = suspicious_function(edge_case_data)
    print(f"✅ Edge case passed: result = {result2}")
except Exception as e:
    print(f"❌ Edge case failed (EXPECTED): {e}")
    debug_logger.error(f"Edge case error: {e}", exc_info=True)
    
    # Detailed traceback
    print("\nDetailed traceback:")
    traceback.print_exc()

# ----------------------------------------
# Step 5 – Deep Dive Analysis
# ----------------------------------------

print("\n" + "-"*50)
print("DEEP DIVE ANALYSIS")
print("-"*50)

# Instrument the function with debugging
def debug_wrapper(func):
    """Wrapper to add debugging to function calls."""
    def wrapper(*args, **kwargs):
        print(f"\n🔍 Calling {func.__name__}")
        print(f"   Args shapes: {[getattr(a, 'shape', type(a)) for a in args]}")
        print(f"   Kwargs: {kwargs}")
        
        try:
            result = func(*args, **kwargs)
            print(f"   ✅ Success! Result shape: {getattr(result, 'shape', type(result))}")
            return result
        except Exception as e:
            print(f"   ❌ Failed with: {e}")
            raise
    return wrapper

# Apply debugging wrapper
suspicious_function_debug = debug_wrapper(suspicious_function)

# Run with debugging
print("\nRunning with debug wrapper:")
try:
    debug_result = suspicious_function_debug(test_data[:5])
except Exception as e:
    print(f"Debug run failed: {e}")

# ----------------------------------------
# Step 6 – Visualization
# ----------------------------------------

print("\n" + "-"*50)
print("VISUALIZATION")
print("-"*50)

# Create debug plots
fig, axes = plt.subplots(2, 2, figsize=(12, 10))
fig.suptitle('Debug Analysis Visualization')

# Plot 1: Input data distribution
ax1 = axes[0, 0]
ax1.hist(test_data.flatten(), bins=50, alpha=0.7)
ax1.set_title('Input Data Distribution')
ax1.set_xlabel('Value')
ax1.set_ylabel('Frequency')

# Plot 2: Data heatmap
ax2 = axes[0, 1]
im = ax2.imshow(test_data[:20, :20], cmap='viridis', aspect='auto')
ax2.set_title('Data Sample Heatmap')
plt.colorbar(im, ax=ax2)

# Plot 3: Error locations (if applicable)
ax3 = axes[1, 0]
# Identify problematic values
problems = np.where(np.isnan(test_data) | np.isinf(test_data))
if len(problems[0]) > 0:
    ax3.scatter(problems[1], problems[0], c='red', s=50)
    ax3.set_title(f'Problem Locations ({len(problems[0])} found)')
else:
    ax3.text(0.5, 0.5, 'No NaN/Inf values found', 
             ha='center', va='center', transform=ax3.transAxes)
    ax3.set_title('Problem Locations')

# Plot 4: Custom analysis
ax4 = axes[1, 1]
# Add custom visualization here
ax4.set_title('Custom Analysis')

plt.tight_layout()
plt.savefig('debug/analysis_visualization.png', dpi=150)
print("✅ Saved visualization to debug/analysis_visualization.png")

# ----------------------------------------
# Step 7 – Findings and Conclusions
# ----------------------------------------

print("\n" + "="*50)
print("FINDINGS AND CONCLUSIONS")
print("="*50)

findings = {
    "timestamp": datetime.now().isoformat(),
    "issue": "Description of the issue",
    "root_cause": "Identified root cause",
    "conditions": [
        "Condition 1 that triggers issue",
        "Condition 2 that triggers issue"
    ],
    "proposed_fix": "Suggested solution",
    "test_results": {
        "normal_case": "passed/failed",
        "edge_case": "passed/failed"
    }
}

print("\n📋 Summary:")
print(f"1. Issue occurs when: {findings['conditions']}")
print(f"2. Root cause: {findings['root_cause']}")
print(f"3. Proposed fix: {findings['proposed_fix']}")

# Save findings
with open('debug/findings.json', 'w') as f:
    json.dump(findings, f, indent=2)
print("\n✅ Findings saved to debug/findings.json")

# ----------------------------------------
# Step 8 – Cleanup
# ----------------------------------------

print("\n" + "-"*50)
print("Debug session completed")
print(f"Log file: debug/debug_log_*.log")
print(f"Plots: debug/analysis_visualization.png")
print(f"Findings: debug/findings.json")
print("-"*50)